clients/src/test/java/org/apache/kafka/test/TestUtils.java

```diff
@@ -247,12 +247,12 @@ public static void waitForCondition(TestCondition testCondition, String conditio
     public static void waitForCondition(TestCondition testCondition, long maxWaitMs, String conditionDetails) throws InterruptedException {
         long startTime = System.currentTimeMillis();
 
-
-        while (!testCondition.conditionMet() && ((System.currentTimeMillis() - startTime) < maxWaitMs)) {
+        boolean conditionMet;
+        while (!(conditionMet = testCondition.conditionMet()) && ((System.currentTimeMillis() - startTime) < maxWaitMs)) {
             Thread.sleep(Math.min(maxWaitMs, 100L));
         }
 
-        if (!testCondition.conditionMet()) {
+        if (!conditionMet) {
```

Comment:
Agreed.

If we did hit this problem, then perhaps we throw an exception if the condition changes between both checks with a message about fixing the condition to make the tests less fragile.